infra/modules/doc-upload-s3/main.tf

```diff
@@ -0,0 +1,55 @@
+data "aws_caller_identity" "current" {}
+data "aws_region" "current" {}
+
+resource "aws_s3_bucket" "doc-upload" {
+  bucket = "${var.environment_name}-document-upload"
+}
+
+resource "aws_s3_bucket_public_access_block" "doc-upload" {
+  bucket = aws_s3_bucket.doc-upload.id
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+resource "aws_s3_bucket_policy" "doc-upload" {
+  bucket = aws_s3_bucket.doc-upload.id
+  policy = data.aws_iam_policy_document.doc-upload.json
+}
```

Comment:
Are we planning on doing the KMS creation and permissioning for encrypting files in another ticket?

Reply:
The `aws_iam_policy_document.doc-upload` policy should have 
```yaml
[
  "kms:Decrypt", 
  "kms:GenerateDataKey"
]
```
permissions in the statement for the KMS arn `aws_kms_key.doc-upload.arn`

The task executor should also have access to the doc-upload policy so that the API can upload files to the bucket

Reply:
To further expand on what @microwavenby said, I think we need:
1. an s3 bucket policy that requires TLS for access
2. an s3 bucket policy that requires the correct KMS key to access
3. an IAM policy to read s3 and decrypt (we should attach this to the `participant` task executor, the `staff` task executor, and the virus scan lambda)
4. an IAM policy to write s3 and decrypt (just the `participant` task executor)
5. an IAM policy to delete s3 and decrypt (just the virus scan lambda)

I think for the S3 bucket policy that requires TLS for access, it should look something like this:

```tf
resource "aws_s3_bucket_policy" "require_tls" {
  bucket = aws_s3_bucket.<bucket>.id
  policy = data.aws_iam_policy_document.require_tls.json
}

data "aws_iam_policy_document" "require_tls" {
  statement {
    sid = "RequireTLS"
    principals {
      type        = "AWS"
      identifiers = ["*"]
    }
    actions = [
      "s3:*",
    ]

    resources = [
      aws_s3_bucket.<bucket>.arn,
      "${aws_s3_bucket.<bucket>.arn}/*"
    ]

    effect = "Deny"

    condition {
      test     = "Bool"
      variable = "aws:SecureTransport"

      values = [
        false
      ]
    }
  }
}
```

and the S3 bucket policy for requiring SSE-KMS encryption, I think should look something like this?

```tf
resource "aws_s3_bucket_policy" "require_kms" {
  bucket = aws_s3_bucket.<bucket>.id
  policy = data.aws_iam_policy_document.require_kms.json
}

data "aws_iam_policy_document" "require_kms" {
  statement {
    sid = "DenyUnEncryptedObjectUploads"
    principals {
      type        = "AWS"
      identifiers = ["*"]
    }
    actions = [
      "s3:PutObject",
    ]

    resources = [
      "${aws_s3_bucket.<bucket>.arn}/*"
    ]

    effect = "Deny"

    condition {
      test     = "StringNotEquals"
      variable = "s3:x-amz-server-side-encryption-aws-kms-key-id"
      values = [
        "${aws_kms_key.<key>.arn}"
      ]
    }
  }
}
```